### PR TITLE
feat(flatMapDeep): Add flatMapDeep

### PR DESCRIPTION
**Pull request by ssi02014**
cloesd: #347 

Add the flatMapDeep discussed in the issue above.

## benchMark

스크린샷 2024-09-01 오후 2 01 18

This is more performant than utilizing lodash or the built-in array functions.

### CODE REVIEW

**Comment by raon0211** [comment]
One question: Would the performance be different between `flatMap(arr, Infinity)` and `flatMapDeep(arr)`?

This is because if possible, I would like to have only one way to implement the same thing. Of course, we might add `flatMapDeep` for compat purposes..

File: docs/reference/array/flatMapDeep.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+It works the same as if you called [Array#flat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat) with [Array#map](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Array/map) as `map(iteratee).flat(Infinity)` in the JavaScript language, but it's faster.
+
+## Signature
+
+```typescript
+// 중첩된 배열을 재귀적으로 풀어내어 가장 내부의 요소 타입을 추출하는 유틸리티 타입이에요.
+type ExtractNestedArrayType<T> = T extends ReadonlyArray<infer U> ? ExtractNestedArrayType<U> : T;
+
+function flattenDeep<T>(arr: readonly T[]): Array<ExtractNestedArrayType<T>>;
```
Comment (raon0211) [inline docs/reference/array/flatMapDeep.md:13]:
```suggestion
function flattenDeep<T>(arr: T[]): Array<ExtractNestedArrayType<T>>;
```

File: docs/ko/reference/array/flatMapDeep.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+JavaScript 언어에 포함된 [Array#flat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat)을 [Array#map](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Array/map)과 함께 `map(iteratee).flat(Infinity)`으로 호출했을 때와 동일하게 동작하지만, 더 빨라요.
+
+## 인터페이스
+
+```typescript
+// 중첩된 배열을 재귀적으로 풀어내어 가장 내부의 요소 타입을 추출하는 유틸리티 타입이에요.
+type ExtractNestedArrayType<T> = T extends ReadonlyArray<infer U> ? ExtractNestedArrayType<U> : T;
+
+function flattenDeep<T>(arr: readonly T[]): Array<ExtractNestedArrayType<T>>;
```
Comment (raon0211) [inline docs/ko/reference/array/flatMapDeep.md:13]:
```suggestion
function flattenDeep<T>(arr: T[]): Array<ExtractNestedArrayType<T>>;
```

**Comment by ssi02014** [comment]
@raon0211 
As a footnote to that function, the return type of flatMap is `Array<FlatArray<U[], D>>`, so specifying the type as `infinity` will not result in the intended type inference.

Therefore, we believe that `flatMapDeep` with the intended type inference is useful and serves the purpose of compatibility with `lodash`.

```ts
const arr = [1];

const newArr1 = flatMapDeep(arr, item => [[[item, item, [[item]]]]]); // number[]

newArr1.map(item => {
  item; // number
  item.toFixed(1); // ✅
});

const newArr2 = flatMap(arr, item => [[[item, item, [[item]]]]], Infinity);
// FlatArray<(number | number[][])[][][], 0 | 1 | -1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19 | 20>[]

newArr2.map(item => {
  item; // FlatArray<(number | number[][])[][][], 0 | 1 | 2 | 3 | 4 | -1 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19 | 20>
 
  item.toFixed(1); // ❌
  (item as number).toFixed(1); // ✅
});
```

스크린샷 2024-09-04 오전 2 01 14

In fact, the performance difference between the two functions is not significant.

**Comment by raon0211** [comment]
Yes, you're right. I accidentally forgot what we discussed 😅

**Review by raon0211 — APPROVED** [review]
Thanks for your great work!

**Pull request merged**